source/mainnet/smart-contracts/guides/unit-test-contract.rst

```diff
@@ -339,18 +339,144 @@ state and balance fields:
        );
        ...
 
-.. warning::
+Reentrancy
+----------
+
+You should watch out for *reentrancy problems*, which can occur when calls to
+``invoke_contract`` end up updating the state of your own contract.
+
+.. code-block:: rust
+
+    let state_copy = *host.state();
+    host.invoke_contract(...);
+
+    // *host.state() and state_copy might not be equal any more due to reentrancy.
+    do_something_with(state_copy);
+
+Another example of reentracny is when the state is *not* updated properly before making an external call.
+This can lead to reentrant calls that pass some validation that is based on the current state, while these calls should fail.
+The classic example of such a security issue is the Ethereum DAO smart contract that was drained of funds due to the reentrancy vulnerability.
+We show a code snippet that implements a small part similar to the DAO contract that stores balances for arbitrary addresses in a map ``StateMap<Address, Amount, S>``.
+The users can request their funds back; if a user is a smart contract, the funds are sent to a specified entrypoint.
```

Comment:
I think this description, and below, is too specific and does not state the key point clearly. 

Re-entrancy is a special case. The point we need to drive across is that invoking another contract passes control out of the current execution context. 

We should also point out that this is not a "novel smart contract issue". It is a standard issue you get when you have, e.g., concurrency. 

So instead of pointing out that they should watch for re-entrancy problems, we should tell them the "positive" thing that they should always make sure to record state updates before doing things that are outside of the contract's control. This way any entrypoint that is ever called will have "clean" state to work with.

We can then list (a simplified) version of the DAO example as a non-example.

Reply:
I think it makes sense to mention that it is a general concurrency issue, as long as we don't write it in a way that requires much knowledge about concurrency by the readers.

Reply:
I'll try to generalise the problem statement and refer to concurrency. I'll make a more detailed description of reentrancy in best practices and cross-reference it with testing.

I've created an issue for adding the full code of the DAO example https://github.com/Concordium/concordium-rust-smart-contracts/issues/215